source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc

```diff
@@ -73,89 +197,165 @@ void Filter::onReadWorker() {
+                     Network::Utility::parseInternetAddressAndPort(
+                         std::string{line_parts[2]} + ":" + std::string{line_parts[4]}),
+                     Network::Utility::parseInternetAddressAndPort(
+                         std::string{line_parts[3]} + ":" + std::string{line_parts[5]})});
     } else if (line_parts[1] == "TCP6") {
-      protocol_version = Network::Address::IpVersion::v6;
-      remote_address = Network::Utility::parseInternetAddressAndPort(
-          "[" + std::string{line_parts[2]} + "]:" + std::string{line_parts[4]});
-      local_address = Network::Utility::parseInternetAddressAndPort(
-          "[" + std::string{line_parts[3]} + "]:" + std::string{line_parts[5]});
+      proxyProtocolHeader_.emplace(
+          WireHeader{0, Network::Address::IpVersion::v6,
+                     Network::Utility::parseInternetAddressAndPort(
+                         "[" + std::string{line_parts[2]} + "]:" + std::string{line_parts[4]}),
+                     Network::Utility::parseInternetAddressAndPort(
+                         "[" + std::string{line_parts[3]} + "]:" + std::string{line_parts[5]})});
     } else {
       throw EnvoyException("failed to read proxy protocol");
     }
+  }
+}
 
-    // Error check the source and destination fields. Most errors are caught by the address
-    // parsing above, but a malformed IPv6 address may combine with a malformed port and parse as
-    // an IPv6 address when parsing for an IPv4 address. Remote address refers to the source
-    // address.
-    const auto remote_version = remote_address->ip()->version();
-    const auto local_version = local_address->ip()->version();
-    if (remote_version != protocol_version || local_version != protocol_version) {
-      throw EnvoyException("failed to read proxy protocol");
+bool Filter::parseExtensions(int fd) {
+  while (proxyProtocolHeader_.value().extensions_length_) {
+    // buf_ is no longer in use so we re-use it to read/discard
+    int bytes_avail;
+    auto& os_syscalls = Api::OsSysCallsSingleton::get();
+    if (os_syscalls.ioctl(fd, FIONREAD, &bytes_avail) < 0) {
+      throw EnvoyException("failed to read proxy protocol (no bytes avail)");
     }
-    // Check that both addresses are valid unicast addresses, as required for TCP
-    if (!remote_address->ip()->isUnicastAddress() || !local_address->ip()->isUnicastAddress()) {
-      throw EnvoyException("failed to read proxy protocol");
+    if (bytes_avail == 0) {
+      return false;
     }
-
-    // Only set the local address if it really changed, and mark it as address being restored.
-    if (*local_address != *socket.localAddress()) {
-      socket.setLocalAddress(local_address, true);
+    bytes_avail = std::min(size_t(bytes_avail), sizeof(buf_));
+    bytes_avail = std::min(size_t(bytes_avail), proxyProtocolHeader_.value().extensions_length_);
+    ssize_t nread = os_syscalls.recv(fd, buf_, bytes_avail, 0);
+    if (nread == -1 && errno == EAGAIN) {
```

Comment:
Same comment about EAGAIN (and elsewhere if applicable. Should not happen, let's fail if it does)